include/ETWProviders/etwprof.h

```diff
@@ -39,32 +39,32 @@ const int kFlagDoubleClick = 100;
 #include <sal.h> // For _Printf_format_string_
 
 // Insert a single event to mark a point in an ETW trace.
-PLATFORM_INTERFACE void ETWMark( const char *pMessage );
+PLATFORM_INTERFACE void ETWMark( _In_z_ PCSTR pMessage );
 // ETWWorkerMark is identical to ETWMark but goes through a different provider,
 // for different grouping.
-PLATFORM_INTERFACE void ETWWorkerMark(const char *pMessage);
+PLATFORM_INTERFACE void ETWWorkerMark(_In_z_ PCSTR pMessage);
 
 // Insert events with one or more generic int or float data fields
-PLATFORM_INTERFACE void ETWMark1I(const char* pMessage, int data1);
-PLATFORM_INTERFACE void ETWMark2I(const char* pMessage, int data1, int data2);
-PLATFORM_INTERFACE void ETWMark1F(const char* pMessage, float data1);
-PLATFORM_INTERFACE void ETWMark2F(const char* pMessage, float data1, float data2);
+PLATFORM_INTERFACE void ETWMark1I(_In_z_ PCSTR pMessage, int data1);
+PLATFORM_INTERFACE void ETWMark2I(_In_z_ PCSTR pMessage, int data1, int data2);
+PLATFORM_INTERFACE void ETWMark1F(_In_z_ PCSTR pMessage, float data1);
+PLATFORM_INTERFACE void ETWMark2F(_In_z_ PCSTR pMessage, float data1, float data2);
 
 // _Printf_format_string_ is used by /analyze
-PLATFORM_INTERFACE void ETWMarkPrintf( _Printf_format_string_ const char *pMessage, ... );
-PLATFORM_INTERFACE void ETWWorkerMarkPrintf( _Printf_format_string_ const char *pMessage, ... );
+PLATFORM_INTERFACE void ETWMarkPrintf( _Printf_format_string_ _In_z_ PCSTR pMessage, ... );

```

Comment:
Nah, it's not possible. The contact for a _Printf_format_string_ is pretty
straightforward -- it behaves like printf. That means it is a
null-terminated string with % specifiers that match the arguments. If that
isn't true then it's not a _Printf_format_string_ but something similar.

I believe that _In_z_ has been combined with _Printf_format_string_
elsewhere, and probably by Microsoft too, but that just means they designed
it poorly, IMHO.

On Fri, May 15, 2015 at 7:04 PM, Alexander Riccio <notifications@github.com>
wrote:

> In include/ETWProviders/etwprof.h
> <https://github.com/google/UIforETW/pull/33#discussion_r30457347>:
>
> >
> >  // _Printf_format_string_ is used by /analyze
> > -PLATFORM_INTERFACE void ETWMarkPrintf( _Printf_format_string_ const char *pMessage, ... );
> > -PLATFORM_INTERFACE void ETWWorkerMarkPrintf( _Printf_format_string_ const char *pMessage, ... );
> > +PLATFORM_INTERFACE void ETWMarkPrintf( _Printf_format_string_ _In_z_ PCSTR pMessage, ... );
>
>  Do you know if it does?
>
> I have no evidence to support any conclusion, but I don't think that it
> does.
>
> If /analyze needs both _Printf_format_string_ and*In_z* then it is being
> very stupid. One should imply the other. I'm not saying it does. I'm just
> saying that it *should*.
>
> I'm *not sure* that it *should*. I wouldn't be surprised if there was
> some absurd use case where a _Printf_format_string_ could also be
> _Out_writes_z_....I'd never advocate writing code like that, but someone
> else *might*.
>
> It seems that other users of _Printf_format_string_ also combine _In_z_
> with _Printf_format_string_
> <https://www.osronline.com/showthread.cfm?link=255907>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/UIforETW/pull/33/files#r30457347>.
>



-- 
Bruce Dawson